q2_metadata/normalization/_norm_errors.py

```diff
@@ -0,0 +1,204 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import yaml
+
+
+class Error(Exception):
+    """Base class for exceptions encountered
+    in the rules checking."""
+
+    @classmethod
+    def generic_message(cls, rule, variable, value):
+        reformatted_value = '\t# %s' % yaml.dump(value).replace('\n', '\n\t# ')
+        return 'Wrong formatting for "%s" rule; variable %s:\n' \
+               '%s' % (rule, variable, reformatted_value)
+
+
+class RuleError(Error):
+    """Exception raised for error due to the
+    presence of a rule that's not recognized."""
+
+    def __init__(self, rule, variable, value):
+        self.rule = rule
+        self.variable = variable
+        self.value = value
+
+    def __str__(self):
+        message = self.generic_message(self.rule, self.variable, self.value)
+        return '%s -> rule not recognized' % message
+
+
+class ExpectedError(Error):
+    """Exception raised for errors in the
+    formatting of the "expected values" rules.
+    """
+    def __init__(self, variable, value):
+        self.rule = 'expected'
+        self.variable = variable
+        self.value = value
+        self.message = self.get_message()
+
+    def get_message(self):
+        if not isinstance(self.value, list):
+            return 'is not a list'
+        str_items = [x for x in self.value if isinstance(x, str)]
```

Comment:
why is there so much type checking?

Reply:
because we can not sanitize humans and that we need to check that the rules are formatted in a way that will allow us to use them, for this one class `ExpectedError()`, if the rule is not giving us a list of strings, but for example, of dict of int -> int, then we just understand what/where are the values that are expected in the current metadata variable.